### ISSUE TITLE
PDP11: timer calibration division by zero

### ISSUE THREAD

**Issue by al20878** [open-simh/simh#245]
- #### Context
PDP11 failed to start, core dumped.  This is a rather rare occurrence.  It happened right after the fresh build from github sources,
at the "post build" register check procedure...  The build was on CYGWIN with gcc-11.

More context:
I was building with `NOVIDEO=1` (and also: `DEBUG=1 NONETWORK=1`), which as it turns out, used to bypass the WINMM high-precision timer on Cygwin (see my PR#242), so it was not using that.  When built with WINMM, I could not reproduce this problem, so it only happened when the timer was "coarse" (default).

coredump should not occur -- division by 0 is very much avoidable if the terms are duly checked.
- #### the output of "sim> SHOW VERSION" while running the simulator which is having the issue
latest build from github as of yesterday
- #### how you built the simulator or that you're using prebuilt binaries

```
LPATH=/usr/local/lib make DEBUG=1 NONETWORK=1 NOVIDEO=1 pdp11
```
- #### the simulator configuration file (or commands) which were used when the problem occurred.

none
- #### the expected behavior and the actual behavior

none
- #### you may also need to provide specific pointers to data files that may be necessary to demonstrate the problem
gdb output pulled post-mortem from the core file:
```
Division by 0:

1168        sim_os_tick_hz = 1000/(sim_os_clock_resoluton_ms * (sim_idle_rate_ms/sim_os_clock_resoluton_ms));
(gdb) p sim_os_tick_hz
$1 = 0
(gdb) p sim_os_clock_resoluton_ms
$2 = 10
(gdb) p sim_idle_rate_ms
$3 = 3
```
Please note the values of the parameters in the formula above (`sim_timer.c`).  Since 'sim_idle_rate_ms / sim_os_clock_resolution_ms` happened to be 0 (the first is smaller than the last), the denominator for division of the 1000 into was 0.  The host machine that was doing the build was quite busy at that time, so maybe that's how the values were "underestimated".  But at any rate, there should be a protection against such thing.

**Comment by markpizz** [comment]
This comes down to if building with cygwin (or MinGW) works for you, great.  If it doesn't you should either build with a free Microsoft Compiler or use pre-built windows binaries.  Cygwin is a Posix emulation on top of Win32 or Win64 and as such may or may not provide precisely correct behavior in various corner cases (such as this).  A well behaved OS syscall layer will never produce the 0 case you are reporting about.

The value of building simh with these tool chains on Windows is essentially 0.

**Comment by al20878** [comment]
This is a runtime issue.  In case of `sim_idle_rate_ms < sim_os_clock_resoluton_ms` you have a problem, and the case goes on unchecked.  It's regardless of what Cygwin is.  You calculate the value of the `idle_rate` with some calibration routine, whose execution timing _depends_ on the load of the CPU.  That value is then used unchecked in the formula.

**Comment by markpizz** [comment]
It certainly is a runtime issue and a problem on cygwin's POSIX emulation.

The problem you are seeing has NEVER occurred on any Linux, macOS or Windows when built with a Microsoft compiler.  This "never" reflects at least the last 12 years, but likely longer.

That it crashes is a bummer, but the only better behavior would be to exit with a message that says something about the less than precise POSIX environment.  If that crash doesn't happen, there is no guarantee that some other subtle hidden issue with the POSIX layer will not cause other issues later with this or other simulators.

**Comment by al20878** [comment]
"NEVER" can happen anytime -- I also saw this once, but luckily I had a debug mode build and a coredump.
If you have a time-dependent code, it's a ticking bomb, which can go off any time, in 12 years or 12 milliseconds.  I guess I had a bad luck.  (Or a good one since I was able to catch it.)  But there *IS* a problem with that code, I'm very certain about that.
```
sim_os_tick_hz = 1000/(sim_os_clock_resoluton_ms * (sim_idle_rate_ms/sim_os_clock_resoluton_ms));
```
I don't exactly know what you are trying to calculate here, but from pure mathematical standpoint, maybe a simpler formula
```
sim_os_tick_hz = 1000 / sim_idle_rate_ms;
```
may work better (or in case when `sim_idle_rate_ms < sim_os_clock_resoluton_ms`)? Just a thought.

I'd say Cygwin just helped to unearth this calculation problem.  There's nothing (un-)POSIX to blame it for.

**Comment by markpizz** [comment]
I have no idea how you got to this "problem".  The line you're referring to is in the middle of this:

```
if ((sim_idle_rate_ms != 0) && (sim_os_clock_resoluton_ms != 0))
    sim_os_tick_hz = 1000/(sim_os_clock_resoluton_ms * (sim_idle_rate_ms/sim_os_clock_resoluton_ms));
else {
    fprintf (stderr, "Can't properly determine host system clock capabilities.\n");
    fprintf (stderr, "Minimum Host Sleep Time:       %u ms\n", sim_os_sleep_min_ms);
    fprintf (stderr, "Minimum Host Sleep Incr Time:  %u ms\n", sim_os_sleep_inc_ms);
    fprintf (stderr, "Host Clock Resolution:         %u ms\n", sim_os_clock_resoluton_ms);
    }
```

Please elaborate how it got a core dump...

**Comment by al20878** [comment]
> Please elaborate how it got a core dump...
I explained that already: "It happened right after the fresh build from github sources,
at the "post build" register check procedure..."
```
BIN/pdp11 RegisterSanityCheck  </dev/null
 Running internal register sanity checks on PDP-11 simulator.
```
at this stage of the build process, of a freshly checked out source clone.

**Comment by markpizz** [comment]
> > Please elaborate how it got a core dump...
>
> I explained that already: "It happened right after the fresh build from github sources,at the "post build" register check procedure..."
> 
> > BIN/pdp11 RegisterSanityCheck  </dev/null
> > Running internal register sanity checks on PDP-11 simulator
>
> .at this stage of the build process, of a freshly checked out source clone.

I phrased the question wrong.  Why do you think the line you mention:
```
    sim_os_tick_hz = 1000/(sim_os_clock_resoluton_ms * (sim_idle_rate_ms/sim_os_clock_resoluton_ms));
```
is the cause of the failure?

If you examine the line of code immediately before the line you've mentioned, if either variables used in the denominator expression of that division has a 0 value, the line you mention won't be executed using a normal compiler on Linux, macOS, and a free Microsoft compiler on Windows.

**Comment by al20878** [comment]
> Why do you think the line you mention:

I don't "think", it's where the program aborted (and dropped a core bomb).  I opened the core in `gdb` and it showed me where that has happened -- that line number.

**Comment by markpizz** [comment]
It seems you've encountered a feature of Cygwin.

Like I said previously, if it works for you on Cygwin, great!  If it doesn't, you should use a Free microsoft compiler or a pre-built Windows binary.

**Comment by markpizz** [comment]
Thanks for pounding on my head until my eyes opened wide enough to see what you were pointing at.

**Comment by al20878** [comment]
I don't think your fix is correct.  If the inequality hold, the division by zero *is to occur*, so you can't proceed, actually
```
if (((sim_idle_rate_ms != 0) && (sim_os_clock_resoluton_ms != 0)) || (sim_idle_rate_ms < sim_os_clock_resoluton_ms))
```
The condition for moving forward with the division has to be ">=".

**Comment by al20878** [comment]
JFTR, the linked PR looks good now. Thanks for the fix

**Comment by markpizz** [comment]
This issue is fixed as of PR #248
